include/exadg/operators/inverse_mass_operator.cpp

```diff
@@ -0,0 +1,364 @@
+  {
+    // In the InverseMassType::MatrixfreeOperator case we can avoid
+    // streaming the vector from memory twice.
+
+    // ghost have to be zeroed out before MatrixFree::cell_loop().
+    dst.zero_out_ghost_values();
+
+    matrix_free->cell_loop(
+      &This::cell_loop_matrix_free_operator,
+      this,
+      dst,
+      src,
+      /*operation before cell operation*/ {}, /*operation after cell operation*/
+      [&](const unsigned int start_range, const unsigned int end_range) {
+        for(unsigned int i = start_range; i < end_range; ++i)
+          dst.local_element(i) *= scaling_factor;
+      },
+      dof_index);
+  }
+  else
+  {
+    apply(dst, src);
+    dst *= scaling_factor;
+  }
+}
+
+template<int dim, int n_components, typename Number>
+void
+InverseMassOperator<dim, n_components, Number>::cell_loop_matrix_free_operator(
+  dealii::MatrixFree<dim, Number> const &,
+  VectorType &       dst,
+  VectorType const & src,
+  Range const &      cell_range) const
+{
+  Integrator                      integrator(*matrix_free, dof_index, quad_index);
+  InverseMassAsMatrixFreeOperator inverse_mass(integrator);
+
+  if(coefficient_is_variable)
+  {
+    for(unsigned int cell = cell_range.first; cell < cell_range.second; ++cell)
+    {
+      integrator.reinit(cell);
+      integrator.read_dof_values(src, 0);
+
+      dealii::AlignedVector<dealii::VectorizedArray<Number>> inverse_JxW_times_coefficient(
+        integrator.n_q_points);
+      inverse_mass.fill_inverse_JxW_values(inverse_JxW_times_coefficient);
+
+      if(consider_inverse_coefficient)
+      {
+        // Consider a mass matrix of the form
+        // (u_h , v_h / c)_Omega
+        // hence fill the vector with (J / c)^-1 = c/J
+        for(unsigned int q = 0; q < integrator.n_q_points; ++q)
+        {
+          inverse_JxW_times_coefficient[q] *=
+            this->variable_coefficients->get_coefficient_cell(cell, q);
+        }
+      }
+      else
+      {
+        // Consider a mass matrix of the form
+        // (u_h , v_h * c)_Omega
+        // hence fill the vector with inv(J * c) = 1/(J * c)
+        for(unsigned int q = 0; q < integrator.n_q_points; ++q)
+        {
+          inverse_JxW_times_coefficient[q] /=
+            this->variable_coefficients->get_coefficient_cell(cell, q);
+        }
+      }
+
+      inverse_mass.apply(inverse_JxW_times_coefficient,
+                         n_components,
+                         integrator.begin_dof_values(),
+                         integrator.begin_dof_values());
+
+      integrator.set_dof_values(dst, 0);
+    }
+  }
+  else
+  {
+    for(unsigned int cell = cell_range.first; cell < cell_range.second; ++cell)
+    {
+      integrator.reinit(cell);
+      integrator.read_dof_values(src, 0);
+
+      inverse_mass.apply(integrator.begin_dof_values(), integrator.begin_dof_values());
+
+      integrator.set_dof_values(dst, 0);
+    }
+  }
+}
+
+template class InverseMassOperator<2, 1, float>;
```

Comment:
we need
scalar version
dim version
dim + 2 version for restart, see
https://github.com/exadg/exadg/blob/213d339016667bff7684ad58f7ea191d9209cf0c/include/exadg/operators/solution_projection_between_triangulations.h#L324C1-L361C6